### PR TITLE
[Gluten-1013] [VL] Support Velox local cache

### PR DESCRIPTION
**Pull request by Yohahaha**
## What changes were proposed in this pull request?

This patch improve Velox local cache support:
- expose more configurations on cache: memory cache size, use o_direct write
- move the velox backend initializations into a separate file

## How was this patch tested?

manual test.

### CODE REVIEW

**Comment by Yohahaha** [comment]
Cloud storage: 10% performance improvement of sf100 with independent session, 20% with single session, both tpcds and tpch.

**Comment by Yohahaha** [comment]
Memory cache does not controlled by Gluten currently, we could use `spark.executor.memoryOverhead` to reserve memory for cache.

AsyncDataCache holds a allocator and control its capacity, allocator can be set explicitly and there is no need to set default allocator.

**Comment by Yohahaha** [comment]
@FelixYBW @zhouyuan 
Could you take a look?

File: cpp/velox/compute/VeloxInitializer.cc [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    std::error_code ec;
+    const std::filesystem::space_info si = std::filesystem::space(ssdCachePathPrefix, ec);
+    if (si.available < ssdCacheSize) {
+      VELOX_FAIL(
+          "not enough space for ssd cache in " + ssdCachePath + " cache size: " + std::to_string(ssdCacheSize) +
+          "free space: " + std::to_string(si.available))
+    }
+
+    velox::memory::MmapAllocator::Options options;
+    options.capacity = memCacheSize;
```
Comment (zhouyuan) [inline cpp/velox/compute/VeloxInitializer.cc:179]: this line will place a limit on the capacity of allocator, so it should also count in `offheap.size`?

**Comment by zhouyuan** [comment]
@Yohahaha thanks for this improvement! got one question on the allocator capacity, the rest changes looks good to me

File: cpp/velox/compute/VeloxInitializer.cc (hunk 1 above)
Comment (FelixYBW) [reply cpp/velox/compute/VeloxInitializer.cc:179]: This part should be counted in offheap.size and spark's storage memory.
@zhztheplayer Do we have interface to report native memory allocation to spark's storage memory?

But the mmapallocator isn't been counted yet,. So let's skip this part for now. It should be counted in executor.overhead now.

Comment (zhztheplayer) [reply cpp/velox/compute/VeloxInitializer.cc:179]: > Do we have interface to report native memory allocation to spark's storage memory?

Not yet, Spark's storage memory API is only for its internal use for now. We can hack that API or just use execution memory without a spiller.

Is the cache going to be reused across Spark tasks? Execution memory is not for that case.

Comment (FelixYBW) [reply cpp/velox/compute/VeloxInitializer.cc:179]: @zhouyuan can the cache be shared in all task thread in the same executor? I'd think so.

@zhztheplayer do we hack the execution memory API?

**Comment by Yohahaha** [comment]
@FelixYBW @zhztheplayer @zhouyuan  Thanks for all replys.

Velox's AsyncDataCache is process-wide, so it can be used across tasks. However, reuse cache in different job need record `mtime`, aka last modify time, Velox does not support mtime yet and don't want to, in my view.

I'm not sure but guess Gluten use off-heap size and manage all native memory under this capacity. If we want use off-heap size as cache capacity, we have to set AsyncDataCache as default allocator. But current implementation provide a pluggable way to use cache and has clearly boundary with current memory management, I think it's enough and better than combine those all now.

Actually, use whole off-heap size as cache capacity is a good solution to leverage fast ssd. We could consider this idea, when Gluten has stable memory management and spill support.

**Comment by FelixYBW** [comment]
Currently offheap size is used by all native memory allocation using malloc, it manage the execution memory. the local cache is more similar to storage memory. Spark's one big feature is the capability to adjust the storage memory and execution memory dynamically. In this way the local cache size can increase or decrease according to Spark's memory management. If we use a memory overhead, its size will be fixed.

but as @zhztheplayer said, currently we still don't have a way to config storage memory, so let temporaly, until we can find a good solution.

File: cpp/velox/compute/VeloxInitializer.cc (hunk 1 above)
Comment (zhztheplayer) [reply cpp/velox/compute/VeloxInitializer.cc:179]: > @zhztheplayer do we hack the execution memory API?

Not it's a public API. We just used it.

Comment (FelixYBW) [reply cpp/velox/compute/VeloxInitializer.cc:179]: Can we submit a PR to Spark community to make storage API public?

Comment (zhztheplayer) [reply cpp/velox/compute/VeloxInitializer.cc:179]: I am not sure, probably Spark has its reasons to make it private. Let's take more look into this part to see what we can do.

**Review by FelixYBW — APPROVED** [review]

**Comment by FelixYBW** [comment]
Thank you @Yohahaha Let's see how much perf it can boost in our nightly test. Should be able to boost several queries even it's local filesystem

**Comment by Yohahaha** [comment]
> Should be able to boost several queries even it's local filesystem

I'm not optimistic about this. In our benchmark on hdfs, it has almost no improvement, same config as below. But looking forward to your result.

> Cloud storage: 10% performance improvement of sf100 with independent session, 20% with single session, both tpcds and tpch. 1G memory cache, 1G ssd cache.

**Review by zhouyuan — APPROVED** [review]
:+1

**Comment by zhouyuan** [comment]
Note: Will update the doc and spark configuration in another PR

**Pull request merged**